### ISSUE TITLE
Audio Output Streams

### ISSUE THREAD

**Issue by mitchmindtree** [MindBuffer/nannou#25]
We could use [cpal](https://docs.rs/cpal) for cross platform audio I/O. Currently it only supports audio output, however I believe they're open to adding support for input - we just need someone to need it enough to actually implement it (and then also implement synchronised duplex streams).

Audio is processed on its own thread and so should likely have its own model->update->render architecture in the same way that the graphics thread has its own model->update->draw.

```rust
mod audio {
    struct Model {}

    fn model(app: &AppAudio) -> Model {}
    fn update(app: &AppAudio, model: Model, event: AudioEvent) -> Model {}
    fn render(app: &AppAudio, model: &Model, buffer: Buffer) -> Buffer {}
}
```

One issue that comes to mind with this structure is that the `render` stage in audio tends to require mutable access to the model in order to step forward the phase of oscillators or step forward a playhead over a buffer of samples. A user could get around this using `Cell` or `RefCell` but this kind of feels like it defeats the purpose.

The `AudioEvent` enum might look something like this:

```rust
enum AudioEvent {
    /// The sample rate of the target `Buffer` has changed.
    SampleRate(f64),
    /// The number of channels per buffer has changed.
    Channels(usize),
    /// The default output device changed.
    DefaultDevice(Device),
    /// A new device was detected.
    DeviceAdded(Device),
    /// A device was removed.
    DeviceRemoved(Device),
}
```

Similarly to `App`, `AppAudio` should act as a general application context but with methods related to audio rather than windowing. Methods might include:

- Device enumeration
- Default device
- Current stream format (sample rate, channels, default sample format / bit depth)

Come to think of it, the device info might even be more useful on the GUI thread, as it is normally via some sort of GUI interaction that devices are viewed or selected.

**Comment by mitchmindtree** [comment]
Another thing worth considering is that, unlike the main graphics thread, it is not uncommon for a user to require multiple audio stream threads at once, occasionally even via different devices.

E.g. the user may want an input stream from one device, process it and sum it onto an output stream via a different device (using some kind of interpolation to synchronise them).

Almost always these streams are interacted with and owned via the main GUI thread, or at least some thread other than the audio streams themselves.

For these reasons it probably makes sense to provide the API for spawning, playing, pausing and dropping streams via the `App` type itself. E.g. when spawning a stream, a `StreamId` or `StreamHandle` can be returned to the user so that they may store it in their `Model` for reference.

**Comment by mitchmindtree** [comment]
Closed by #26.
